raft.go

```diff
@@ -649,6 +655,29 @@ func (r *raft) maybeSendAppend(to uint64, sendIfEmpty bool) bool {
 	return true
 }
 
+func (r *raft) maybeSendEmptyAppend(to uint64, pr *tracker.Progress) bool {
+	// TODO(pav-kv): when pr.Next is updated, we always know the term of entry
+	// pr.Next-1, because the append message contains it. Store (Next-1, Term) in
+	// Progress, instead of just Next. Then we don't have to fetch the term and
+	// send a potentially unnecessary snapshot here.
+	prevTerm, err := r.raftLog.term(pr.Next - 1)
+	if err != nil {
+		// The log probably got truncated at >= pr.Next, so we can't catch up the
+		// follower log anymore. Send a snapshot instead.
+		return r.maybeSendSnapshot(to, pr)
+	}
+	r.send(pb.Message{
+		To:      to,
+		Type:    pb.MsgApp,
+		Index:   pr.Next - 1,
+		LogTerm: prevTerm,
+		Commit:  r.raftLog.committed,
```

Comment:
I'm noticing the odd asymmetry between `MsgApp` sending the latest Commit index and the follower computing `min(committed, lastnewi)`, and `MsgHeartbeat` carefully sending `min(pr.Match, r.raftLog.committed)` to avoid a follower-side panic. Do you have thoughts about that?

Reply:
Yes. The commit index in `MsgApp` is "optimistic", and can be larger than `pr.Match`. As a result, a follower can simultaneously append and commit entries in some cases. Technically, `Commit` index can even exceed the last index of the entries in the batch. The way I think about `MsgApp` is that it's effectively 2 independent messages:
1. A [logSlice](https://github.com/etcd-io/raft/blob/83d8deca057fef78cf3a86ea70457e4d3255e4da/types.go#L38-L67) of the leader log to append to the follower log.
2. A `(term, commit)` is the point in the leader@`term` log that is committed (see #144 explaining the "leader log" coordinate system).

As such, even if the append (1) is redundant/no-op (because the follower already has these entries from elsewhere), the commit index (2) can still communicate valuable information - if the follower log is at the same term, it can advance the commit index safely. We don't take advantage of this currently (the commit index is [advanced](https://github.com/etcd-io/raft/blob/83d8deca057fef78cf3a86ea70457e4d3255e4da/log.go#L127) only within the bounds of the appended entries (1)), but we should.

The commit index in `MsgHeartbeat` is "pessimistic" and cut at `pr.Match` because currently the follower has a [panic](https://github.com/etcd-io/raft/blob/83d8deca057fef78cf3a86ea70457e4d3255e4da/raft.go#L1802) if the index is not in the bounds of the follower log. This panic is unnecessary though, see #138 and #139 aiming to relax it.

I would like to clean-up the semantics of the `Commit` indices in both messages, to both be treated as (2) above - a `(term, index)` that the leader sends without relying on the knowledge about the follower log. The follower log has enough knowledge to apply it safely.